source/pki_utils.c

```diff
@@ -226,3 +226,76 @@ int aws_read_and_decode_pem_file_to_buffer_list(
 
     return AWS_OP_SUCCESS;
 }
+
+int aws_decode_pem_to_der_buf(
+    struct aws_allocator *alloc,
+    const struct aws_byte_cursor *pem_cursor,
+    struct aws_byte_buf *der_buf) {
+    aws_byte_buf_clean_up(der_buf);
+    struct aws_array_list base_64_buffer_list;
+
+    if (aws_array_list_init_dynamic(&base_64_buffer_list, alloc, 2, sizeof(struct aws_byte_buf))) {
+        return AWS_OP_ERR;
+    }
+
+    int err_code = AWS_OP_ERR;
+
+    if (s_convert_pem_to_raw_base64(alloc, pem_cursor, &base_64_buffer_list)) {
+        goto cleanup_base64_buffer_list;
+    }
+
+    for (size_t i = 0; i < aws_array_list_length(&base_64_buffer_list); ++i) {
+        size_t decoded_len = 0;
+        struct aws_byte_buf *byte_buf_ptr = NULL;
+        aws_array_list_get_at_ptr(&base_64_buffer_list, (void **)&byte_buf_ptr, i);
+        struct aws_byte_cursor byte_cur = aws_byte_cursor_from_buf(byte_buf_ptr);
+
+        if (aws_base64_compute_decoded_len(&byte_cur, &decoded_len)) {
+            aws_raise_error(AWS_IO_FILE_VALIDATION_FAILURE);
+            goto cleanup_output_due_to_error;
+        }
+
+        struct aws_byte_buf decoded_buffer;
+
+        if (aws_byte_buf_init(&decoded_buffer, alloc, decoded_len)) {
+            goto cleanup_output_due_to_error;
+        }
+
+        if (aws_base64_decode(&byte_cur, &decoded_buffer)) {

```

Comment:
this might be easy to optimize. length of decoded is guaranteed to be smaller than length of encoded. so we can allocated based on encoded length. we can probably also pull allocation out of the loop